tests/e2e/graceful_shutdown_test.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+	"go.etcd.io/etcd/tests/v3/framework/interfaces"
+)
+
+func TestGracefulShutdown(t *testing.T) {
+	clusterSize := 3
+	testRunner := e2e.NewE2eRunner()
+	testRunner.BeforeTest(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	clus := testRunner.NewCluster(ctx, t, config.WithClusterSize(clusterSize))
+	// clean up orphaned resources like closing member client.
+	defer clus.Close()
+	// shutdown each etcd member process sequentially
+	// and start from old leader, (new leader), (follower)
+	tryShutdownLeader(ctx, t, clus.Members())
+}
+
+// tryShutdownLeader tries stop etcd member if it is leader.
+// it also asserts stop leader should not take longer than 1.5 seconds and leaderID has been changed within 500ms.
+func tryShutdownLeader(ctx context.Context, t *testing.T, members []interfaces.Member) {
+	quorum := len(members)/2 + 1
+	for len(members) >= quorum {
```

Comment:
```suggestion
	for len(members) > quorum {
```

When `len(members) == quorum`, there is no way to automatically elect a new leader after stopping current leader. If I understand it correctly, the reason why all workflow checks pass is that etcd closes the peer connections after it stops current member. But it's too fragile, because even the current member(leader) successfully transfers the leadership to another member, but the new leader may also lose the leadership in the 500ms sleep.

https://github.com/etcd-io/etcd/blob/b7ad0382ad983c46a5c06f1015531d1ad5bfce70/server/embed/etcd.go#L442-L454

Reply:
> When len(members) == quorum, there is no way to automatically elect a new leader after stopping current leader. 

https://github.com/etcd-io/etcd/blob/b7ad0382ad983c46a5c06f1015531d1ad5bfce70/server/etcdserver/server.go#L1185-L1192 guarantees leadership transfer must complete before `s.HardStop()`. I intentionally wants to add coverage on that. 

> If I understand it correctly, the reason why all workflow checks pass is that etcd closes the peer connections after it stops current member. 

status() API does not require active quorum. Even if the old leader was shutting down, the remaining etcd server can still response to client without any issues. The failed CI test was failing because of unrelated test case. `2023-05-09T03:03:58.9456980Z --- FAIL: TestV3MetricsSecure (6.58s)` 

I've locally tested with `-count=100` with no failures.

Reply:
`Status` doesn't depend on quorum, but it might return a `raft.None` leader, and accordingly fail the case.

Will the case fail if you change the sleep (line 65) to a value >2s? My point is it might not be good to depend on the WALL time.

Reply:
> Will the case fail if you change the sleep (line 65) to a value >2s?

No, it won't. I just tested locally with `-count=100`. 

<img width="844" alt="image" src="https://github.com/etcd-io/etcd/assets/54131596/785b50f1-fbb6-4fd1-bf9c-37925a063256">

Proposal 1 and 2 makes sense to me.

Reply:
I've made the suggested change except externalizing and customizing the election timeout, closing this thread~

Reply:
> Will the case fail if you change the sleep (line 65) to a value >2s? 

Please try a value > 2s, e.g. 3s or 4s.

Reply:
Yeah, just tested again locally. 

<img width="1163" alt="image" src="https://github.com/etcd-io/etcd/assets/54131596/2ec6c8b8-acf2-4be4-919f-dbe5dc488816">

Reply:
Regardless, I've made the suggested change `len(members) > quorum` since there is a concern of `len(members) >= quorum` will make this test flaky.

Reply:
Please see https://github.com/etcd-io/etcd/pull/15828#discussion_r1189188620